src/Features/Core/Portable/CodeRefactorings/SyncNamespace/AbstractSyncNamespaceCodeRefactoringProvider.RenameNamespaceCodeAction.cs

```diff
@@ -0,0 +1,295 @@
+                //
+                // And for documents referencing those symbols, they can be linked/MTFM documents themselves.
+                // In which case, we treat them as independent documents and fix them individually, and at the 
+                // end rely on existing conflict resolving mechanism to merge those fixes in the same file together.
+
+                foreach (var declaredSymbol in declaredSymbols)
+                {
+                    var refSymbols = await SymbolFinder.FindReferencesAsync(
+                      declaredSymbol,
+                      solution,
+                      cancellationToken).ConfigureAwait(false);
+
+                    var refLocationsForSymbol = refSymbols.Where(refSymbol => refSymbol.Definition == declaredSymbol).SelectMany(refSymbol => refSymbol.Locations);
+
+                    // Ignore other documents with identical file path as triggering document (i.e. other TFM in a MTFM project)
+                    refLocationsInCurrentDocument = refLocationsInCurrentDocument.Concat(refLocationsForSymbol.Where(loc => loc.Document.Id == document.Id));
+                    refLocationsInOtherFilePath = refLocationsInOtherFilePath.Concat(refLocationsForSymbol.Where(loc => !PathUtilities.PathsEqual(loc.Document.FilePath, document.FilePath)));
+                }   
+
+                // Transform current document(s) first
+                solution = (await FixDeclarationDocumentAsync(document, refLocationsInCurrentDocument, cancellationToken).ConfigureAwait(false)).Project.Solution;
+
+                // Then fix all referencing documents.
+                var oldImport = _service.CreateUsingDirective(OldNamespaceParts);
+                var newImport = _service.CreateUsingDirective(NewNamespaceParts);
+
+                var refLocationGroups = refLocationsInOtherFilePath.GroupBy(loc => loc.Document.Id);
+                foreach (var refInOneDocument in refLocationGroups)
+                {
+                    var refDocument = solution.GetDocument(refInOneDocument.Key);
+                    solution = (await FixReferencingDocumentAsync(refDocument, oldImport, newImport, refInOneDocument, cancellationToken).ConfigureAwait(false)).Project.Solution;
+                }
+
+                return ImmutableArray.Create<CodeActionOperation>(new ApplyChangesOperation(solution));
+            }            
+
+            private async Task<Document> FixDeclarationDocumentAsync(Document document, IEnumerable<ReferenceLocation> refLocations, CancellationToken cancellationToken)
+            {
+                Debug.Assert(!NewNamespaceParts.IsDefault);
+
+                // 1. Reduce all references.
+                // 2. Add usings for containing namespaces, in case we have references relying on old namespace declaration for resolution. 
+                // 3. Change namespace declaration.
+                // 4. Remove unnecessary usings.              
```

Comment:
these qre quite cryptic.  it's fine to actualy explain what each of these means, and why you want to do it.